formulaic/materializers/narwhals.py

```diff
@@ -0,0 +1,216 @@
+
+import functools
+import itertools
+from typing import TYPE_CHECKING, Any, Dict, List, Sequence, Tuple, cast
+
+import narwhals.dtypes as nwd
+import narwhals.stable.v1 as nw
+import numpy
+import pandas
+import scipy.sparse as spsparse
+from interface_meta import override
+
+from formulaic.utils.cast import as_columns
+from formulaic.utils.null_handling import drop_rows as drop_nulls
+
+from .base import FormulaMaterializer
+
+if TYPE_CHECKING:  # pragma: no cover
+    from formulaic.model_spec import ModelSpec
+
+
+class NarwhalsMaterializer(FormulaMaterializer):
+    REGISTER_NAME = "narwhals"
+    REGISTER_INPUTS: Sequence[str] = ("pandas.core.frame.DataFrame", "pandas.DataFrame")
+    REGISTER_OUTPUTS: Sequence[str] = ("narwhals", "pandas", "numpy", "sparse")
+
+    @override
+    def _init(self) -> None:
+        self.__narwhals_data = nw.from_native(self.data)
+        self.__data_context = self.__narwhals_data.to_dict()
+
+    @override  # type: ignore
+    @property
+    def data_context(self):
+        return self.__data_context
+
+    @override
+    def _is_categorical(self, values: Any) -> bool:
+        if isinstance(values, nw.Series):
+            if not isinstance(values.dtype, nwd.NumericType):
+                return True
+        return super()._is_categorical(values)
+
+    @override
+    def _encode_constant(
+        self,
+        value: Any,
+        metadata: Any,
+        encoder_state: Dict[str, Any],
+        spec: ModelSpec,
+        drop_rows: Sequence[int],
+    ) -> Any:
+        nrows = self.nrows - len(drop_rows)
+        if spec.output == "sparse":
+            return spsparse.csc_matrix(numpy.array([value] * nrows).reshape((nrows, 1)))
+        series = value * numpy.ones(nrows)
+        return series
+
+    @override
+    def _encode_numerical(
+        self,
+        values: Any,
+        metadata: Any,
+        encoder_state: Dict[str, Any],
+        spec: ModelSpec,
+        drop_rows: Sequence[int],
+    ) -> Any:
+        if drop_rows:
+            values = drop_nulls(values, indices=drop_rows)
+        if spec.output == "sparse":
+            return spsparse.csc_matrix(
+                numpy.array(values).reshape((values.shape[0], 1))
+            )
+        return values
+
+    @override
+    def _encode_categorical(
+        self,
+        values: Any,
+        metadata: Any,
+        encoder_state: Dict[str, Any],
+        spec: ModelSpec,
+        drop_rows: Sequence[int],
+        reduced_rank: bool = False,
+    ) -> Any:
+        # Even though we could reduce rank here, we do not, so that the same
+        # encoding can be cached for both reduced and unreduced rank. The
+        # rank will be reduced in the _encode_evaled_factor method.
+        from formulaic.transforms import encode_contrasts
+
+        if drop_rows:
+            values = drop_nulls(values, indices=drop_rows)
+        if isinstance(values, nw.Series):
+            values = values.to_pandas()
```

Comment:
I think it should be possible to keep this Narwhals-native, as we have:
- `cat.get_categories`: https://narwhals-dev.github.io/narwhals/api-reference/expr_cat/#narwhals.expr.ExprCatNamespace.get_categories
- `Series.to_dummies`: https://narwhals-dev.github.io/narwhals/api-reference/series/#narwhals.series.Series.to_dummies

will try this out and let you know